Benchmarking with different Motion Planners (STOMP, CHOMP, OMPL)

raghavendersahdev opened the pull request:
### Benchmarking with different Motion Planners (STOMP, CHOMP, OMPL)

This PR is a work done by Raghavender as a part of 2018 Google summer of code. As a part of this PR, benchmarking works for different motion planners (CHOMP, STOMP, OMPL) available in MoveIt. For a particular benchmark(start sates, planning scene, goal states, queries, etc.), different planners can be evaluated.

### Checklist
- [ ] addition of yaml files and launch files for including STOMP and CHOMP planners in environments with and without obstacles.
- [ ] fixing the existing benchmarking code to make it work with STOMP and CHOMP.

mamoll commented on moveit_ros/benchmarks/examples/demo_obstacles.yaml:
```diff
@@ -0,0 +1,33 @@
+        - plugin: chomp_interface/CHOMPPlanner
+          planners:
+            - CHOMP
+        - plugin: stomp_moveit/StompPlannerManager
```
maybe `stomp_moveit` should be renamed to `stomp_interface` to be consistent with OMPL and CHOMP?

davetcoleman replied:
I'm guessing this is tied by the package name

raghavendersahdev replied:
yes this is the package name: `stomp_moveit`

mamoll replied:
My point was that perhaps the package should be renamed for consistency. This can be done later, if others agree.

davetcoleman replied:
i agree it would be nice

davetcoleman commented on moveit_ros/benchmarks/examples/demo2.yaml:
```diff
@@ -27,3 +27,4 @@ benchmark_config:
         - plugin: my_plugin_ns/my_plugin
           planners:
             - MyPlanner
+
```
please don't introduce changes like this into PRs since nothing else in this file was modified

raghavendersahdev replied:
sure this was my mistake, removed this change

davetcoleman commented on moveit_ros/benchmarks/src/BenchmarkExecutor.cpp:
```diff
@@ -532,6 +532,9 @@ bool BenchmarkExecutor::plannerConfigurationsExist(const std::map<std::string, s
     planning_interface::PlannerManagerPtr pm = planner_interfaces_[it->first];
     const planning_interface::PlannerConfigurationMap& config_map = pm->getPlannerConfigurations();
 
+    if (pm->getDescription().compare("stomp") || pm->getDescription().compare("chomp"))
```
what is this for? add comment inline

raghavendersahdev replied:
added comments explaining this, I put this here because for STOMP and CHOMP, the benchmarking was throwing an error as [this line](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/benchmarks/src/BenchmarkExecutor.cpp#L533) for CHOMP/STOMP returns a size of 0 for the config_map parameter which was causing an error. After putting this check, everything worked fine.

mamoll commented on moveit_ros/benchmarks/src/BenchmarkExecutor.cpp:
```diff
@@ -532,6 +532,11 @@ bool BenchmarkExecutor::plannerConfigurationsExist(const std::map<std::string, s
     const planning_interface::PlannerConfigurationMap& config_map = pm->getPlannerConfigurations();
 
+    // if the planner is chomp or stomp skip this function and return true for checking planner configurations for the
+    // planning group other wise an error occurs, this is kind of a trick which works well
```
Even with `other wise` replaced with `otherwise`, I still can't parse this.  Why is this a trick? Perhaps the issue that with OMPL, a specific planning algorithm needs to be defined for a planning group, whereas with STOMP and CHOMP this is not necessary.

raghavendersahdev replied:
changed comment

raghavendersahdev commented:
Sorry, I did something bad here while rebasing, accidentally rebased with the wrong local branch (raghavendersahdev/moveit kinetic branch) rather than the actual remote ros-planning/moveit kinetic branch, I am fixing it right now.

davetcoleman commented:
its not clear if i should review this

raghavendersahdev commented:
@davetcoleman I had a question for you, is it possible to undo a rebase after I already pushed it (I think it is kind of complicated) ? 

I actually wanted to rebase my [`raghavendersahdev/moveit benchmarking branch`](https://github.com/raghavendersahdev/moveit/tree/benchmarking) with the [`ros-planning/moveit kinetic branch`](https://github.com/ros-planning/moveit/tree/kinetic-devel) ...But accidentally I rebased this with my own current [`raghavendersahdev/moveit kinetic branch`](https://github.com/raghavendersahdev/moveit/tree/kinetic-devel).....So what ended up happening was => the wrong rebase merged the commits from my kinetic branch which also has a [PR#987 open](https://github.com/ros-planning/moveit/pull/987) currently......

This specifc PR has only 6 commits but the branch which I used to rebase this with had 10 commits so I ended up with 16 commits in this PR. 

Should I get rid of this PR and open a new one ? .... I am sorry for this confusion pertaining to this specific PR.

davetcoleman commented:
you can branch off of ``kinetic-devel`` then cherry-pick the relevant commits back on top of that, and it should work out

raghavendersahdev commented:
ok, thanks Dave, I will read up on `git cherry-pick` and try to get rid of the 10 non-relevant commits in my [`raghavendersahdev/moveit benchmarking` ](https://github.com/raghavendersahdev/moveit/tree/benchmarking) branch to only have the original 6 commits in the benchmarking branch

raghavendersahdev commented:
Hi @davetcoleman , I have fixed this PR and removed the redundant commits added due to my wrong rebase earlier. This PR is now ready for review and Travis passes.

davetcoleman commented on moveit_ros/benchmarks/examples/demo_obstacles.yaml:
```diff
@@ -0,0 +1,33 @@
+# of motion plan queries and start states in the Kitchen scene.
+
+# Five planners from two different plugins are run a total of 50 times each, with a
+# maximum of 10 seconds per run.  Output is stored in the /home/benchmarks directory.
```
> Output is stored in the /home/benchmarks directory.
     
This is not true, right? Line 18 has a different location

raghavendersahdev replied:
yes, you are right, fixed!

davetcoleman commented on moveit_ros/benchmarks/examples/demo_panda_all_planners.yaml:
```diff
@@ -0,0 +1,33 @@
+# of motion plan queries and start states in the Kitchen scene.
+
+# Five planners from two different plugins are run a total of 50 times each, with a
+# maximum of 10 seconds per run.  Output is stored in the /home/benchmarks directory.
```
same directory issue

raghavendersahdev replied:
fixed

davetcoleman commented on moveit_ros/benchmarks/src/BenchmarkExecutor.cpp:
```diff
@@ -532,6 +532,12 @@ bool BenchmarkExecutor::plannerConfigurationsExist(const std::map<std::string, s
+    // planning group otherwise an error occurs, because for OMPL a specific planning algorithm needs to be defined for
+    // a planning group, whereas with STOMP and CHOMP this is not necessary
+    if (pm->getDescription().compare("stomp") || pm->getDescription().compare("chomp"))
+      return true;
```
this should be ``continue`` not ``return``... in case OMPL is also in the list of planner interfaces to check

raghavendersahdev replied:
fixed

davetcoleman commented:
cherry picked to melodic
